lib/Basics/ThreadGuard.h

```diff
@@ -0,0 +1,54 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <thread>
+#include <vector>
+
+//
+struct ThreadGuard {
+  ThreadGuard() = default;
+  explicit ThreadGuard(size_t reserve) : threads(reserve) {}
+  ~ThreadGuard() {
+    joinAll();  // Should we really?
+  }
+
+  template<typename Function, typename... Args>
+  auto emplace(Function&& f, Args... args) -> void {
+    threads.emplace(std::forward<Function>(f), std::forward<Args...>(args...));
+  }
+
+  void joinAll() {
+    // TODO: Copypasta
+    for (auto& t : threads) {
+      if (t.joinable()) {
+        try {
+          t.join();
+        } catch (...) {
+        }
```

Comment:
Do we really want to silently swallow all exceptions? IMHO join should normally never throw, but if it does we would want to know!

Reply:
I only copied this code from the original because I wanted to have exactly this discussion;

If we do not `try-catch(...)`, and call `joinAll` in the destructor, it could throw; arguably, if the `joinAll()` throws, everything is in a bad state in any case so maybe we accept that our destructor throws